src/System.Private.CoreLib/shared/System/IO/Stream.cs

```diff
@@ -189,6 +189,72 @@ private int GetCopyBufferSize()
             return bufferSize;
         }
 
+        public virtual void CopyTo(ReadOnlySpanAction<byte, object?> callback, object? state, int bufferSize)
+        {
+            if (callback == null) throw new ArgumentNullException(nameof(callback));
+
+            CopyTo(new WriteCallbackStream(callback, state), bufferSize);
+        }
+
+        public virtual Task CopyToAsync(Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask> callback, object? state, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (callback == null) throw new ArgumentNullException(nameof(callback));
+
+            return CopyToAsync(new WriteCallbackStream(callback, state), bufferSize, cancellationToken);
+        }
+
+        private sealed class WriteCallbackStream : Stream
+        {
+            private readonly ReadOnlySpanAction<byte, object?>? _action;
+            private readonly Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask>? _func;
+            private readonly object? _state;
+
+            public WriteCallbackStream(ReadOnlySpanAction<byte, object?> action, object? state)
+            {
+                _action = action;
+                _state = state;
+            }
+
+            public WriteCallbackStream(Func<ReadOnlyMemory<byte>, object?, CancellationToken, ValueTask> func, object? state)
+            {
+                _func = func;
+                _state = state;
+            }
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                Write(new ReadOnlySpan<byte>(buffer, offset, count));
+            }
+
+            public override void Write(ReadOnlySpan<byte> span)
+            {
+                if (_action == null) throw new NotSupportedException();
```

Comment:
Probably shouldn't throw in this situation because a derived `Stream` could have overridden the existing `CopyTo/CopyToAsync` overloads with code that mismatches sync/async calls.

Reply:
https://github.com/dotnet/coreclr/pull/27639#discussion_r341830598